Support different isolation levels

yingjunwu opened the pull request:
In this PR, I attempt to introduce different isolation levels, i.e., `READ_ONLY`, `SERIALIZABLE`, `REPEATABLE_READS`, `READ_COMMITTED`, into Peloton.

Please note that, for now, our `SERIALIZABLE` is equivalent to `REPEATABLE_READS`, as we do not solve phantom problem. We do not support `READ_UNCOMMITTED` as it is too trival. Following Oracle's [article](http://www.oracle.com/technetwork/testcontent/o65asktom-082389.html), I treat `READ_ONLY` as a special isolation level.

I think @jessesleeping is the best person for reviewing this PR, but as he's enjoying life at Facebook, we need to find another person to review it.

jessesleeping commented on src/concurrency/local_epoch.cpp:
```diff
@@ -16,17 +16,29 @@
+    
+      if (is_snapshot_read == true) {
+    
+        epoch_lower_bound_ = epoch_id - 1;
```
Why do we set the epoch_lower_bound_ to a smaller epoch? I think it's supposed to be a non-decreasing number, otherwise the GC process will be confused.

jessesleeping replied:
@yingjunwu ?

yingjunwu replied:
@jessesleeping The local thread is unaware of the progress of any other threads. It is possible that thread `TH1` is already at epoch 100, but another thread `TH2` is still at epoch 10. In this case, the centralized epoch manager should tell `TH1` that the progress from a global view is still at epoch 10.

jessesleeping commented on src/concurrency/timestamp_ordering_transaction_manager.cpp:
```diff
@@ -397,83 +163,188 @@ bool TimestampOrderingTransactionManager::AcquireOwnership(
 // will release the write lock anyway.
 void TimestampOrderingTransactionManager::YieldOwnership(
-    UNUSED_ATTRIBUTE Transaction *const current_txn, const oid_t &tile_group_id,
+    UNUSED_ATTRIBUTE Transaction *const current_txn, 
```
Remove the UNUSED_ATTRIBUTE tag. The argument is actually used.

yingjunwu replied:
it's used in assertion. the compiler will complain when compiling under release mode.

jessesleeping commented on src/concurrency/decentralized_epoch_manager.cpp:
```diff
@@ -18,33 +18,34 @@ namespace concurrency {
 
   // enter epoch with thread id
-  cid_t DecentralizedEpochManager::EnterEpoch(const size_t thread_id) {
+  cid_t DecentralizedEpochManager::EnterEpoch(const size_t thread_id, const bool is_snapshot_read) {
```
Does is_snapshot_read == true mean ``SNAPSHOT`` or ``READONLY``?

yingjunwu replied:
yes.

jessesleeping replied:
Which one? ``SNAPSHOT ISOLATION`` or ``READONLY``?

yingjunwu replied:
both. Whenever a transaction is executed under snapshot isolation or declared as read-only, it will directly access a snapshot version.

jessesleeping commented on src/concurrency/timestamp_ordering_transaction_manager.cpp:
```diff
@@ -397,83 +163,188 @@ bool TimestampOrderingTransactionManager::AcquireOwnership(
+    } else {
+      // if it's not select for update, then directly return true.
+      // no need to update read/write set.
+      return true;
```
Why don't we update the last read timestamp here? What if the read version is write-locked by a transaction that has a timestamp earlier than the reader?

yingjunwu replied:
because it's for snapshot isolation. Under this isolation level, a transaction can read a version without setting the last read timestamp.

jessesleeping replied:
Let's say we have the following txn:
- txnA (ts = 6)
- txnB (ts = 5)

And we have the following tuples/versions:
- tuple0_v0 (begin_ts = 3, end_ts = inf)
- tuple1_v0 (begin_ts = 4, end ts = inf)

Now we have the following operations:
1. txnA reads tuple0 -> txnA read tuple0_v0
2. txnB write locks tuple1
3. txnA reads tuple1 -> txnA read tuple1_v0
4. txnB updates tuple1 -> tuple1_v0 (begin_ts = 4, end_ts = 5), tuple1_v1 (begin_ts = 5, end_ts = inf)
5. txnB commits
6. txnA commits

In this case, A will end up reading tuple0_v0 and tuple1_v0. But for a consistent snapshot, it should read tuple0_v0 and tuple1_v1.

jessesleeping replied:
@yingjunwu ?

jessesleeping replied:
Never mind.. I think that's an anomaly.

yingjunwu replied:
Let me think about this.

yingjunwu replied:
Please read the original snapshot isolation paper https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tr-95-51.pdf first, and then we discuss offline.

yingjunwu commented:
This PR requires careful code review. In addition to implementing all isolation levels, I refactored the epoch manager code to make it easier to understand. I removed all the old logging code and prepared APIs for the new logging component. I commented out the storage manager code because it is tightly coupled with the old logging code. According to our plan, we will refactor the storage management component in May. As discussed before, I also temporarily commented out some logging-related test cases.

jessesleeping commented:
My dear Yingjun, how can I review a PR of +3,912/−9,540.....

jessesleeping commented:
Please add more test cases for different isolation levels.

yingjunwu commented:
@jessesleeping I think you could skip the logging part, because it's just some function declarations. You may start from reviewing epoch manager as well as the GC manager. I refactored them and unified API. I think we need to rewrite all the related test cases as the current test cases do not really test different isolation levels...

jessesleeping commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -50,18 +52,19 @@ void TransactionLevelGCManager::Running(const int &thread_id) {
     // When the DBMS has started working but it never processes any transaction,
-    // we may see max_cid == MAX_CID.
-    if (max_cid == MAX_CID) {
+    // we may see expired_eid == MAX_EID.
```
Please explain why we will observe that equation.

yingjunwu replied:
I think I've already explained it. Initially, `expired_eid` is set to `MAX_EID`. So it is possible for the GC to observe the equation.

jessesleeping commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -81,13 +84,13 @@ void TransactionLevelGCManager::Running(const int &thread_id) {
 }
 
 
-void TransactionLevelGCManager::RecycleTransaction(std::shared_ptr<GCSet> gc_set, const cid_t &timestamp) {
```
Why did we use timestamp to hash the context? I think I may be the one who wrote this.. but I cannot remember why I did that..

yingjunwu replied:
You were reviewing the removed code. Originally we use timestamp to hash the context. But now I have changed it to `thread_id`. We no longer pass timestamp to the GC, because the GC only needs to know the `epoch_id` of a GC context.

jessesleeping commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -135,27 +138,28 @@ int TransactionLevelGCManager::Unlink(const int &thread_id, const cid_t &max_cid
   }  // end for
 
-  auto safe_max_cid = concurrency::TransactionManagerFactory::GetInstance().GetNextCommitId();
+  eid_t safe_expired_eid = concurrency::EpochManagerFactory::GetInstance().GetNextEpochId();
```
May be we should add a COMPILER FENCE before acquiring this epoch id? We need to reach a barrier here.

yingjunwu replied:
why? The only requirement is that `safe_expired_eid` is larger than the current `epoch_id`. Even if the compiler/hardware reorders the instruction, it does not affect the correctness of the GC.

yingjunwu commented:
@jessesleeping I agree with you that we need to spend more time testing the system backend, including transaction manager, epoch manager, GC, and logging. It's not that easy and we should review the code carefully.

apavlo commented:
Disable logger :-1:

yingjunwu commented:
@jessesleeping I have finished implementing `REPEATABLE_READS` and `SNAPSHOT_ISOLATION`. Please perform another round of code review. Thanks.

chenboy commented on src/concurrency/timestamp_ordering_transaction_manager.cpp:
```diff
@@ -397,83 +184,286 @@ bool TimestampOrderingTransactionManager::AcquireOwnership(
+    if (acquire_ownership == true) {
+
+      // get the latest version of this tuple.
+      location = *(tile_group_header->GetIndirection(location.offset));
```
I don't quite understand here. why can we acquire ownership of a version that is newer than the snapshot version?

yingjunwu replied:
it's `select_for_update`. We read the snapshot version and update the latest version. We need to acquire write locks to prevent write-write conflicts. See Section VIII.B in the paper: Multi-version Concurrency via Timestamp Range Conflict Management.

chenboy commented on src/executor/update_executor.cpp:
```diff
@@ -139,44 +139,67 @@ bool UpdateExecutor::DExecute() {
+    ///////////////////////////////////////////////////////////
+    // if running at snapshot isolation, 
+    // then we need to retrieve the latest version of this tuple.
+    if (current_txn->GetIsolationLevel() == IsolationLevelType::SNAPSHOT) {
```
Same here. I don't understand why can we update a version that is newer than the snapshot version.

yingjunwu replied:
Similarly, because we always need to update the latest version of a tuple, we can hold a lock on a version that is newer than the snapshot one.

Another paper for your reference: a critique of ansi sql isolation levels.

yingjunwu commented:
@apavlo I've talked to @pmenon before making this change. The reason is that the original design of storage manager is heavily dependent on the logging component, which is unreasonable. I think that the storage manager requires a complete redesign. As `sorter` only needs memory, we can safely replace it with malloc and free.

apavlo commented:
Only the core parts of the storage system should be calling malloc/free directly. I don't see why this change is necessary.

yingjunwu commented:
@apavlo I understand. But anyway we have to completely refactor the storage manager code. It depends on logging stuff, which is wired. Meanwhile, it also assumes the underlying NVRAM is an emulated one, which is unreasonable for a real-world DBMS. I think we can have a discussion on the storage manager part after SIGMOD.

yingjunwu commented on src/concurrency/timestamp_ordering_transaction_manager.cpp:
```diff
@@ -84,280 +83,64 @@ void TimestampOrderingTransactionManager::InitTupleReserved(
-         tile_group_header->GetBeginCommitId(tuple_id) == MAX_CID;
+  auto tuple_begin_cid = tile_group_header->GetBeginCommitId(tuple_id);
+  
+  return tuple_begin_cid == MAX_CID;
```
done.

yingjunwu commented on src/concurrency/timestamp_ordering_transaction_manager.cpp:
```diff
@@ -397,83 +184,286 @@ bool TimestampOrderingTransactionManager::AcquireOwnership(
+  } // end SNAPSHOT
+
+  //////////////////////////////////////////////////////////
+  //// handle READ_COMMITTED
```
I think we'd better keep them in separate so as to make the logic clean enough for review.

yingjunwu commented on src/concurrency/transaction.cpp:
```diff
@@ -122,16 +116,17 @@ void Transaction::RecordUpdate(const ItemPointer &location) {
       return;
     }
     PL_ASSERT(false);
+  } else {
```
It's possible. Just consider the `select_for_update` case. The DBMS locks the version without recording it into read/write set.

Comments added.

yingjunwu commented on src/concurrency/transaction.cpp:
```diff
@@ -172,7 +165,8 @@ bool Transaction::RecordDelete(const ItemPointer &location) {
     PL_ASSERT(false);
   } else {
-    PL_ASSERT(false);
+    rw_set_[tile_group_id][tuple_id] = RWType::UPDATE;
```
good catch here. should change `RWType::UPDATE` to `RWType::DELETE`. Similar as the previous one, a DBMS can delete a tuple after it `select_for_update` this tuple.

yingjunwu commented:
@apavlo So I have revised the code based on the review comments. The snapshot isolation part is done. However, the logging component is still unfinished, as it depends on the PR #577. Also, the storage management component should be redesigned.

apavlo commented:
@yingjunwu Let's discuss in person. I need to figure out what PRs we are going to need to merge.

yingjunwu commented:
I think once @seojungmin fixed the `optimizer_sql_test` bug, the PR can be merged.

haoxianghua commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -96,18 +100,19 @@ int TransactionLevelGCManager::Unlink(const int &thread_id,
-        }
-        return res;
-      });
+    [this, &garbages, &tuple_counter, expired_eid]
```
'this' value is never used in the lambda function

yingjunwu replied:
done.

Michael-Tieying-Zhang commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -96,18 +100,19 @@ int TransactionLevelGCManager::Unlink(const int &thread_id,
 
   // First iterate the local unlink queue
   local_unlink_queues_[thread_id].remove_if(
```
The function for `local_unlink_queues_` is that if an item in `unlink_queue_` is not garbage for now, that item should be put back again for the future gc. The reason you separate `local_unlink_queues_` from `unlink_queue_` is that `unlink_queue_` uses `LockFreeQueue`, but `local_unlink_queues_` uses `std::List`. So why `unlink_queue_` can not use  `std::List` as well, then we don't need `local_unlink_queues_`?

yingjunwu replied:
`LockFreeQueue` is thread-safe, while `List` is not.

haoxianghua commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -116,14 +121,14 @@ int TransactionLevelGCManager::Unlink(const int &thread_id,
       break;
```
In this part, we have unlink_queues_ and local_unlink_queues_. Is that right because unlink_queues_ uses peloton::LockFreeQueue and this queue doesn't support push the unrequired garbage_ctx back, so you maintain local_unlink_queues using std::list to save unrequired garbage_ctx. If so, can I use std::list instead? Because every thread has its own queue so we don't need to consider concurrency issue.

yingjunwu replied:
`local_unlink_queue` is thread-local and will be updated by a single GC thread, while `unlink_queue` is not because worker thread can push to-be-recycled tuple versions into it.

yingjunwu commented on src/include/concurrency/decentralized_epoch_manager.h:
```diff
@@ -86,30 +100,29 @@ class DecentralizedEpochManager : public EpochManager {
+
 
   inline uint32_t GetNextTransactionId() {
     return next_txn_id_.fetch_add(1, std::memory_order_relaxed);
```
I need Matt's PR #577 to fully decentralize the transaction id generation.

yingjunwu commented on src/include/gc/transaction_level_gc_manager.h:
```diff
@@ -34,20 +34,20 @@ namespace gc {
 
 
 struct GarbageContext {
```
There's no functon in this struct.

yingjunwu commented on src/include/gc/transaction_level_gc_manager.h:
```diff
@@ -91,7 +91,7 @@ class TransactionLevelGCManager : public GCManager {
     this->is_running_ = false;
   }
 
```
done.

yingjunwu commented on src/concurrency/decentralized_epoch_manager.cpp:
```diff
@@ -18,69 +18,70 @@ namespace concurrency {
-      }
-    }
-  }
+      return (snapshot_global_epoch_id_ << 32) | 0x0;
```
Yes. They are snapshot transactions. So we can assign them with same commit id.

yingjunwu commented on src/concurrency/local_epoch.cpp:
```diff
@@ -16,61 +16,53 @@
-    auto epoch_map_itr = epoch_map_.find(epoch_id);
-    // check whether the corresponding epoch exists.
-    if (epoch_map_itr == epoch_map_.end()) {
+    epoch_lock_.Lock();
```
The global epoch manager thread needs to read data from LocalEpoch. So we must add lock to guarantee atomicity.

yingjunwu commented on src/concurrency/local_epoch.cpp:
```diff
@@ -16,61 +16,53 @@
-      epoch_map_[epoch_id] = epoch_ptr;
+      epoch_id_lower_bound_ = epoch_id - 1;
+    
+    } else if (epoch_id_lower_bound_ >= epoch_id) {
```
The local epoch progress can be faster than the global one. In this case, `epoch_id_lower_bound_` can be larger than `epoch_id`.

yingjunwu commented on src/concurrency/local_epoch.cpp:
```diff
@@ -82,35 +74,37 @@ namespace concurrency {
-        PL_ASSERT(epoch_ptr->epoch_id_ > epoch_lower_bound_);
-        epoch_lower_bound_ = epoch_ptr->epoch_id_ - 1;
+        PL_ASSERT(epoch_ptr->epoch_id_ > epoch_id_lower_bound_);
+        epoch_id_lower_bound_ = epoch_ptr->epoch_id_ - 1;
```
The global epoch manger should also be responsible for updating local epoch. Otherwise, if there's no transaction assigned to a certain thread, its corresponding epoch manager will never be updated.

yingjunwu replied:
The global epoch manager can update the lower bound.

yingjunwu commented on src/concurrency/local_epoch.cpp:
```diff
@@ -82,35 +74,37 @@ namespace concurrency {
+        // get the smallest epoch.
         auto &epoch_ptr = epoch_queue_.top();
+        // if there's no transaction in the epoch
         if (epoch_ptr->txn_count_ == 0) {
```
The global epoch manger should also be responsible for updating local epoch. Otherwise, if there's no transaction assigned to a certain thread, its corresponding epoch manager will never be updated.

Michael-Tieying-Zhang replied:
> The global epoch manger should also be responsible for updating local epoch.


Updating `txn_count_`? Where is that?

yingjunwu replied:
While it is true that we do not update  `txn_count_` in `GetExpiredEpochId()`, I suggest leaving the code segment there as we may further optimize the code to remove the `epoch_lock_`. In that case data race can occur and we need to check `txn_count_` in `GetExpiredEpochId()`.

domino-succ replied:
Ok. Please add TODO tag in the code.

yingjunwu commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -134,29 +139,31 @@ int TransactionLevelGCManager::Unlink(const int &thread_id,
+  eid_t safe_expired_eid = concurrency::EpochManagerFactory::GetInstance().GetCurrentEpochId();
+
+  for(auto& item : garbages){
+      reclaim_maps_[thread_id].insert(std::make_pair(safe_expired_eid, item));
```
It's a multi-map.

Michael-Tieying-Zhang replied:
`std::multimap` is ordered. I guess you mean `std::unordered_multimap`.

yingjunwu replied:
We should not use `std::unordered_multimap`. Our design is to sort the garbage versions according to epoch id. Versions with the oldest epoch id can be reclaimed first. Please check the code in `Reclaim()` (https://github.com/yingjunwu/peloton/blob/f5ec56f1c7f6e95f6e19a2409857f05e7ed6bb61/src/gc/transaction_level_gc_manager.cpp#L155). I have comments there.

apavlo commented:
Other than @Michael-Tieying-Zhang's [comment](https://github.com/cmu-db/peloton/pull/585#discussion_r120016263), is everything else ready to go in this PR?

If so, then @yingjunwu can rebase and I will merge.

yingjunwu commented:
@apavlo I suggest we should perform a performance test before or after merging the PR. Now we can support different isolation levels. We can test whether varying the isolation levels help improve the performance.

ChTimTsubasa commented:
I would do a test no later than tomorrow as soon as I got the account.

apavlo commented:
@yingjunwu I don't think we should worry about performance at this point. We should worry about correctness and stability. Can you rebase?

@ChTimTsubasa What test are you referring to? If you mean OLTP-Bench, then it will crash because of the bug that's fixed in #672. We can't run OLTP-Bench without that.

yingjunwu commented:
Please refer to PR #691.
